File: app/services/ai/feedback_processor.rb
module Ai
  class FeedbackProcessor
    SYSTEM_PROMPT = <<~PROMPT
      You are an AI assistant that analyzes customer feedback. For each feedback item, you must:
      1. Categorize it as one of: bug, feature_request, complaint, or uncategorized
      2. Assign a priority: p1 (critical/urgent), p2 (high), p3 (medium), p4 (low)
      3. Provide a brief 1-2 sentence summary
      4. Rate your confidence from 0.0 to 1.0

      Respond in JSON format only:
      {
        "category": "bug|feature_request|complaint|uncategorized",
        "priority": "p1|p2|p3|p4",
        "summary": "Brief summary here",
        "confidence": 0.85
      }

      Priority Guidelines:
      - P1: System down, security issues, data loss, blocking many users
      - P2: Major functionality broken, significant user impact
      - P3: Minor bugs, usability issues, non-critical features
      - P4: Nice-to-haves, minor improvements, cosmetic issues
    PROMPT

    def initialize(project: nil, include_objectives: true)
      api_key = Integration.anthropic_api_key(project: project)
      @client = Anthropic::Client.new(api_key: api_key)
      @project = project
      @include_objectives = include_objectives
    end

    def process(feedback, objectives: nil)
      return if feedback.ai_processed_at.present? && !feedback.manually_reviewed

      prompt = build_prompt(feedback)

      begin
        response = @client.messages.create(
          model: "claude-sonnet-4-20250514",
          max_tokens: 500,
          system: SYSTEM_PROMPT,
          messages: [ { role: "user", content: prompt } ]
        )

        result = parse_response(response)

        # Optionally analyze objective alignment
        alignment_result = nil
        if @include_objectives
          objectives ||= load_active_objectives(feedback)
          if objectives.any?
            alignment_result = analyze_objective_alignment(feedback, objectives)
          end
        end

        update_feedback(feedback, result, alignment_result)

        { success: true, result: result, alignment: alignment_result }
      rescue => e
        handle_error(feedback, e)
        { success: false, error: e.message }
      end
    end

    def process_batch(feedbacks)
      results = { processed: 0, failed: 0, errors: [] }

      feedbacks.each do |feedback|
        result = process(feedback)
        if result[:success]
          results[:processed] += 1
        else
          results[:failed] += 1
          results[:errors] << { id: feedback.id, error: result[:error] }
        end

        # Rate limiting - be nice to the API
        sleep(0.5)
      end

      results
    end

    private

    def build_prompt(feedback)
      parts = []
      parts << "Title: #{feedback.title}" if feedback.title.present?
      parts << "Content: #{feedback.content}"
      parts << "Source: #{feedback.source}"
      parts << "Author: #{feedback.author_name}" if feedback.author_name.present?

      parts.join("\n\n")
    end

    def parse_response(response)
      content = response.content&.first&.text
      raise JSON::ParserError, "No content in response" unless content

      # Extract JSON from response (handle potential markdown code blocks)
      json_match = content.match(/\{.*\}/m)
      raise JSON::ParserError, "No JSON found in response" unless json_match

      JSON.parse(json_match[0])
    end

    def update_feedback(feedback, result, alignment_result = nil)
      attrs = {
        category: result["category"],
        priority: result["priority"],
        ai_summary: result["summary"],
        ai_confidence_score: result["confidence"].to_f,
        ai_processed_at: Time.current
      }

      if alignment_result
        attrs[:objective_alignment_score] = alignment_result[:alignment_score]
        attrs[:aligned_objective_ids] = alignment_result[:aligned_objective_ids]
      end

      feedback.update!(attrs)
    end

    def load_active_objectives(feedback)
      return [] unless feedback.project

      BusinessObjective.for_ai_context(project: feedback.project)
    end

    def analyze_objective_alignment(feedback, objectives)
      aligner = ObjectiveAligner.new(project: @project)
      aligner.analyze_feedback(feedback, objectives: objectives)
    rescue => e
      Rails.logger.error("Objective alignment failed for Feedback##{feedback.id}: #{e.message}")
      nil
    end

    def handle_error(feedback, error)
      Rails.logger.error("AI processing failed for Feedback##{feedback.id}: #{error.message}")
      feedback.update!(
        category: :uncategorized,
        priority: :unset,
        ai_summary: "AI processing failed: #{error.message}",
        ai_confidence_score: 0.0,
        ai_processed_at: Time.current
      )
    end

    def handle_parse_error(feedback, error)
      Rails.logger.error("AI response parsing failed for Feedback##{feedback.id}: #{error.message}")
      feedback.update!(
        category: :uncategorized,
        priority: :unset,
        ai_summary: "AI response could not be parsed",
        ai_confidence_score: 0.0,
        ai_processed_at: Time.current
      )
    end
  end
end

File: apps/worker/src/ai/feedback-processor.ts
/**
 * AI feedback classification — ports Rails FeedbackProcessor.
 * Categorizes, prioritizes, and summarizes feedback items.
 */
import { eq, isNull } from "drizzle-orm";
import type { Database } from "@customer-pulse/db/client";
import { feedbacks, projects } from "@customer-pulse/db/client";
import { callClaudeJson, resolveApiKey } from "./call-claude.js";

interface ClassificationResult {
  category: string;
  priority: string;
  summary: string;
  confidence_score: number;
  objective_alignment_score?: number;
  aligned_objective_ids?: number[];
}

const CATEGORY_MAP: Record<string, number> = { uncategorized: 0, bug: 1, feature_request: 2, complaint: 3 };
const PRIORITY_MAP: Record<string, number> = { unset: 0, p1: 1, p2: 2, p3: 3, p4: 4 };

const SYSTEM_PROMPT = `You are a product feedback classifier. Analyze customer feedback and return a JSON object with:
- "category": one of "bug", "feature_request", "complaint", "uncategorized"
- "priority": one of "p1" (critical/blocking), "p2" (high/important), "p3" (medium/nice-to-have), "p4" (low/minor)
- "summary": 2-3 sentence summary for a product manager
- "confidence_score": 0.0-1.0 how confident you are in the classification

Respond with ONLY the JSON object, no markdown fences.`;

const SYSTEM_PROMPT_WITH_OBJECTIVES = `You are a product feedback classifier. Analyze customer feedback and return a JSON object with:
- "category": one of "bug", "feature_request", "complaint", "uncategorized"
- "priority": one of "p1" (critical/blocking), "p2" (high/important), "p3" (medium/nice-to-have), "p4" (low/minor)
- "summary": 2-3 sentence summary for a product manager
- "confidence_score": 0.0-1.0 how confident you are in the classification
- "objective_alignment_score": 0.0-1.0 how aligned this feedback is with the business objectives (1.0 = directly supports critical objective)
- "aligned_objective_ids": array of objective IDs this feedback relates to (empty array if none)

Respond with ONLY the JSON object, no markdown fences.`;

export async function processFeedbackItem(
  db: Database,
  feedbackId: number,
): Promise<boolean> {
  const [row] = await db.select().from(feedbacks).where(eq(feedbacks.id, feedbackId)).limit(1);
  if (!row) return false;

  // Check for business objectives context
  const [proj] = await db
    .select({ objectives: projects.businessObjectives })
    .from(projects)
    .where(eq(projects.id, row.projectId))
    .limit(1);

  const hasObjectives = !!proj?.objectives?.trim();

  const userContent = hasObjectives
    ? `Business Objectives:\n${proj!.objectives}\n\nFeedback:\nTitle: ${row.title ?? "(none)"}\n\n${row.content}`
    : `Title: ${row.title ?? "(none)"}\n\n${row.content}`;

  const result = await callClaudeJson<ClassificationResult>({
    system: hasObjectives ? SYSTEM_PROMPT_WITH_OBJECTIVES : SYSTEM_PROMPT,
    user: userContent,
    maxTokens: 512,
  });

  const now = new Date();

  if (!result) {
    // Fallback: just mark as processed with defaults
    await db
      .update(feedbacks)
      .set({
        aiSummary: "AI classification unavailable — ANTHROPIC_API_KEY may not be set.",
        aiProcessedAt: now,
        aiConfidenceScore: 0,
        updatedAt: now,
      })
      .where(eq(feedbacks.id, feedbackId));
    return false;
  }

  await db
    .update(feedbacks)
    .set({
      category: CATEGORY_MAP[result.category] ?? 0,
      priority: PRIORITY_MAP[result.priority] ?? 0,
      aiSummary: result.summary ?? "No summary generated.",
      aiProcessedAt: now,
      aiConfidenceScore: result.confidence_score ?? 0.5,
      updatedAt: now,
    })
    .where(eq(feedbacks.id, feedbackId));

  return true;
}

export async function processFeedbackBatch(
  db: Database,
  batchSize: number = 100,
): Promise<{ processed: number; remaining: boolean }> {
  const apiKey = await resolveApiKey();
  if (!apiKey) {
    console.warn("[ai] ProcessFeedbackBatch skipped — no Anthropic API key configured");
    return { processed: 0, remaining: false };
  }

  // Fetch unprocessed feedbacks
  const unprocessed = await db
    .select({ id: feedbacks.id })
    .from(feedbacks)
    .where(isNull(feedbacks.aiProcessedAt))
    .limit(batchSize);

  let processed = 0;
  for (const row of unprocessed) {
    const ok = await processFeedbackItem(db, row.id);
    if (ok) processed++;
  }

  // Check if there are more
  const [remaining] = await db
    .select({ id: feedbacks.id })
    .from(feedbacks)
    .where(isNull(feedbacks.aiProcessedAt))
    .limit(1);

  return { processed, remaining: !!remaining };
}
